### PR TITLE
Accept any &dyn Array in nullif kernel

### LINKED ISSUE #510: `nullif` operating on `ArrayRef`
**Is your feature request related to a problem or challenge? Please describe what you are trying to do.**

The `nullif` kernel currently only applies to `PrimitiveArray<T>`. This makes it impossible to use with non-primitive types, which often arise when working with structured / nested data. Additionally, it seems like the `nullif` kernel should have a similar signature to `filter` -- where `filter` selects only the non-true rows,  `nullif` nulls out the true rows.

**Describe the solution you'd like**
The signature of `nullif` changed to `nullif(values: ArrayRef, condition: BooleanArray) -> Result<ArrayRef>`.
Possibly the addition of the complement `null_unless`.

I have a draft that I think would work (can open a PR), but wanted to make sure that such a kernel is desirable, and understand whether it should use a different name to avoid API breakages (`generic_nullif` or something like that).

### PR DESCRIPTION
**Pull request by tustvold**
# Which issue does this PR close?

Closes #510

# Rationale for this change
 

# What changes are included in this PR?

# Are there any user-facing changes?

### CODE REVIEW

File: arrow/src/compute/kernels/boolean.rs [hunk 1]
```diff
@@ -1136,6 +1135,278 @@ mod tests {
...
+
+        let s = s.slice(2, 3);
+        let select = select.slice(1, 3);
+        let select = as_boolean_array(select.as_ref());
+        let a = nullif(s.as_ref(), select).unwrap();
+        let r: Vec<_> = as_string_array(&a).iter().collect();
+        assert_eq!(r, vec![None, Some("a"), None]);
+    }
+
+    #[test]
```
Comment (tustvold) [inline arrow/src/compute/kernels/boolean.rs:1178]: These tests are borrowed from @bjchambers original PR #521

File: arrow-buffer/src/buffer/ops.rs [hunk 2]
```diff
@@ -66,10 +66,10 @@ pub fn bitwise_bin_op_helper<F>(
     right: &Buffer,
     right_offset_in_bits: usize,
     len_in_bits: usize,
-    op: F,
+    mut op: F,
 ) -> Buffer
 where
-    F: Fn(u64, u64) -> u64,
+    F: FnMut(u64, u64) -> u64,
```
Comment (tustvold) [inline arrow-buffer/src/buffer/ops.rs:72]: This is just a drive by fix to make this accept mutable closures - this is a backwards compatible change

**Comment by alamb** [comment]
cc @jhorstmann

**Comment by bjchambers** [comment]
Didn't examine it, but looks a lot simpler than the previous attempts, so that's nice! Curious if something changed to make this easier?

**Comment by tustvold** [comment]
> Curious if something changed to make this easier

I don't think anything changed, but rather than offsetting values, it pads the null buffer. This is simpler and highly unlikely to have a meaningful performance or memory impact, as zero-allocating a buffer is effectively free, and null masks are very compressed

File: arrow/src/compute/kernels/boolean.rs [hunk 3]
```diff
@@ -470,27 +473,19 @@ pub fn is_not_null(input: &dyn Array) -> Result<BooleanArray> {
 
 /// Copies original array, setting null bit to true if a secondary comparison boolean array is set to true.
```
Comment (viirya) [inline arrow/src/compute/kernels/boolean.rs:474]: Not related to the change, but I got confused while reading this.

Based on following comment:
```
    // left=1          right=1 & comp=true    output bitmap=null
    // left=1          right=1 & comp=false   output bitmap=set
```

I think that if the boolean array is set to true, null bit will be set to false?

**Review by viirya — APPROVED** [review]

File: arrow/src/compute/kernels/boolean.rs [hunk 4]
```diff
@@ -499,69 +494,72 @@ where
...
+                let t = l & !r;
+                valid_count += t.count_ones() as usize;
+                t
+            })
+        }
+        None => bitwise_unary_op_helper(&right, right_offset, len, |b| {
+            let t = !b;
+            valid_count += t.count_ones() as usize;
+            t
+        }),
```
Comment (viirya) [inline arrow/src/compute/kernels/boolean.rs:527]: The logic looks the same as before.

File: arrow/src/compute/kernels/boolean.rs (hunk 3 above)
Comment (tustvold) [reply arrow/src/compute/kernels/boolean.rs:474]: I think this is confusion caused by the fact the null masks, are actually validity masks... Will clarify docs

**Pull request merged**